ops/testing.py

```diff
@@ -14,3823 +14,52 @@
-                last_data=data or {},
-                repeat_after=repeat_after,
-            )
-            self._notices[unique_key] = notice
-
-        return notice.id, new_or_repeated
-
-    def _api_error(self, code: int, message: str) -> pebble.APIError:
-        status = http.HTTPStatus(code).phrase
-        body = {
-            'type': 'error',
-            'status-code': code,
-            'status': status,
-            'result': {'message': message},
-        }
-        return pebble.APIError(body, code, status, message)
-
-    def get_notice(self, id: str) -> pebble.Notice:
-        for notice in self._notices.values():
-            if notice.id == id:
-                return notice
-        raise self._api_error(404, f'cannot find notice with ID "{id}"')
-
-    def get_notices(
-        self,
-        *,
-        users: Optional[pebble.NoticesUsers] = None,
-        user_id: Optional[int] = None,
-        types: Optional[Iterable[Union[pebble.NoticeType, str]]] = None,
-        keys: Optional[Iterable[str]] = None,
-    ) -> List[pebble.Notice]:
-        # Similar logic as api_notices.go:v1GetNotices in Pebble.
-
-        filter_user_id = 0  # default is to filter by request UID (root)
-        if user_id is not None:
-            filter_user_id = user_id
-        if users is not None:
-            if user_id is not None:
-                raise self._api_error(400, 'cannot use both "users" and "user_id"')
-            filter_user_id = None
-
-        if types is not None:
-            types = {(t.value if isinstance(t, pebble.NoticeType) else t) for t in types}
-        if keys is not None:
-            keys = set(keys)
-
-        notices = [
-            notice
-            for notice in self._notices.values()
-            if self._notice_matches(notice, filter_user_id, types, keys)
-        ]
-        notices.sort(key=lambda notice: notice.last_repeated)
-        return notices
-
-    @staticmethod
-    def _notice_matches(
-        notice: pebble.Notice,
-        user_id: Optional[int] = None,
-        types: Optional[Set[str]] = None,
-        keys: Optional[Set[str]] = None,
-    ) -> bool:
-        # Same logic as NoticeFilter.matches in Pebble.
-        # For example: if user_id filter is set and it doesn't match, return False.
-        if user_id is not None and not (notice.user_id is None or user_id == notice.user_id):
-            return False
-        if types is not None and notice.type not in types:
-            return False
-        if keys is not None and notice.key not in keys:
-            return False
-        return True
+__all__ = [
+    # Expose the Harness functionality in this namespace.
+    'ActionFailed',
+    'ActionOutput',
+    'AppUnitOrName',
+    # These names are exposed for backwards compatibility.
+    'CharmBase',
+    'CharmMeta',
+    'CharmType',
+    'Container',
+    'ExecArgs',
+    'ExecHandler',
+    'ExecProcess',
+    'ExecResult',
+    'Harness',
```

Comment:
Presumably `Harness` should be in the top group, in other words, the "Expose the Harness functionality" group. I wonder if these groups got a bit messed up -- it seems there should be a lot more in that group.

Reply:
Yeah, the auto-sorting breaks things if it's in one list, so I have it in two now. I put the PR back into draft until things were resolved - I wasn't expecting reviews until I un-drafted it.

Reply:
Oops -- my bad. I was just going through my email notifications and missed that it was a draft.